### PR TITLE
return string for rails 4 compatibility

### PR DESCRIPTION
**Pull request by scottbarrow**
* 	return string for rails 4 compatibility

### CODE REVIEW

File: lib/ransack_ui/ransack_overrides/helpers/form_builder.rb [hunk 1]
```diff
@@ -143,7 +143,7 @@ def attribute_collection_for_base(base, attributes=nil)
           if c == 'id'
             foreign_klass = object.context.traverse(base).model_name
             # Check that model can autocomplete. If not, skip this id column.
-            next nil unless foreign_klass.constantize._ransack_autocompletes_through.present?
+            next nil unless foreign_klass.to_s.constantize._ransack_autocompletes_through.present?
```
Comment (avit) [inline lib/ransack_ui/ransack_overrides/helpers/form_builder.rb]: what is this even doing... 
`.model_name` and then `.constantize`? Why not just use the `foreign_klass` directly?

File: lib/ransack_ui/ransack_overrides/helpers/form_builder.rb [hunk 2]
```diff
@@ -167,7 +167,7 @@ def attribute_collection_for_base(base, attributes=nil)
             }.to_json
           end
 
-          if foreign_klass && (autocomplete_source = foreign_klass.constantize._ransack_autocompletes_through rescue false)
+          if foreign_klass && (autocomplete_source = foreign_klass.to_s.constantize._ransack_autocompletes_through rescue false)
```
Comment (avit) [inline lib/ransack_ui/ransack_overrides/helpers/form_builder.rb]: Same here... could we just remove `.to_s.constantize`?

File: lib/ransack_ui/ransack_overrides/helpers/form_builder.rb (hunk 1 above)
Comment (scottbarrow) [reply lib/ransack_ui/ransack_overrides/helpers/form_builder.rb]: ah good point

File: lib/ransack_ui/ransack_overrides/helpers/form_builder.rb [hunk 3]
```diff
@@ -167,7 +167,7 @@ def attribute_collection_for_base(base, attributes=nil)
             }.to_json
           end
 
-          if foreign_klass && (autocomplete_source = foreign_klass.constantize._ransack_autocompletes_through rescue false)
+          if foreign_klass && (autocomplete_source = foreign_klass._ransack_autocompletes_through rescue false)
```
Comment (avit) [inline lib/ransack_ui/ransack_overrides/helpers/form_builder.rb]: Maybe pull this long `()` into its own line please, it's a bit much for an inline assignment on `if`:

`if autocomplete_source` ...

rescue should be `nil` too (like a `try`).

File: lib/ransack_ui/ransack_overrides/helpers/form_builder.rb [hunk 4]
```diff
@@ -141,9 +141,9 @@ def attribute_collection_for_base(base, attributes=nil)
 
           # Set model name as label for 'id' column on that model's table.
           if c == 'id'
-            foreign_klass = object.context.traverse(base).model_name
+            foreign_klass = object.context.traverse(base)
             # Check that model can autocomplete. If not, skip this id column.
-            next nil unless foreign_klass.constantize._ransack_autocompletes_through.present?
+            next nil unless foreign_klass._ransack_autocompletes_through.present?
```
Comment (avit) [inline lib/ransack_ui/ransack_overrides/helpers/form_builder.rb]: Autocomplete should be `nil` when not set, so shouldn't need to also call `.present?`

**Review by avit — APPROVED** [review]
Looks like it'll work as expected, just some general comments.

File: lib/ransack_ui/ransack_overrides/helpers/form_builder.rb [hunk 5]
```diff
@@ -198,6 +198,12 @@ def attribute_collection_for_base(base, attributes=nil)
       rescue UntraversableAssociationError => e
         nil
       end
+
+      private
+
+      def autocomplete_source
+        @autocomplete_source ||= @foreign_klass && @foreign_klass._ransack_autocompletes_through
```
Comment (avit) [inline lib/ransack_ui/ransack_overrides/helpers/form_builder.rb]: I'm not sure if the instance variable for `@foreign_klass` is going to be a safe approach.

The method `def attribute_collection_for_base(base, attributes=nil)` takes a base object: if this is different between method calls and we use it to redefine this then things could get mixed up?

**Review by avit — CHANGES_REQUESTED** [review]

File: lib/ransack_ui/ransack_overrides/helpers/form_builder.rb (hunk 5 above)
Comment (scottbarrow) [reply lib/ransack_ui/ransack_overrides/helpers/form_builder.rb]: right, I've added a rescue, just to keep the `_ransack_autocompletes_through` extracted but not dependent, this is pretty much what it was doing previously

**Review by PMK9 — APPROVED** [review]

**Pull request merged**